### PR TITLE
Humanize array ser/deser

### PR DESCRIPTION
**Pull request by al8n**
This PR solves the serialization/deserialization problem in OpenAPI, which was mentioned in Discord.

### CODE REVIEW

File: consensus-engine/src/types.rs [hunk 1]
```diff
@@ -190,6 +190,81 @@ impl Qc {
...
+                match (l.to_digit(16), r.to_digit(16)) {
+                    (Some(l), Some(r)) => *byte = (l as u8) << 4 | r as u8,
+                    (_, _) => {
+                        return Err(<D::Error as serde::de::Error>::custom(format!(
+                            "Invalid character pair '{l}{r}'"
+                        )))
+                    }
+                };
+            }
+            Ok(output)
```
Comment (danielSanchezQ) [inline consensus-engine/src/types.rs:250]: This should be a single function.

File: consensus-engine/src/types.rs [hunk 2]
```diff
@@ -190,6 +190,81 @@ impl Qc {
...
+                    Err(<D::Error as serde::de::Error>::invalid_length(
+                        bytes.len(),
+                        &format!("{N}").as_str(),
+                    ))
+                } else {
+                    let mut output = [0u8; N];
+                    output.copy_from_slice(bytes);
+                    Ok(output)
+                }
+            })
```
Comment (danielSanchezQ) [inline consensus-engine/src/types.rs:263]: This as well a single function

File: consensus-engine/src/types.rs [hunk 3]
```diff
@@ -190,6 +190,81 @@ impl Qc {
     }
 }
 
+mod util {
```
Comment (danielSanchezQ) [inline consensus-engine/src/types.rs:193]: Maybe we can have a utils module, inside a serde one:
`consensus-engine/utils/serde`?

File: consensus-engine/src/types.rs [hunk 4]
```diff
@@ -190,6 +190,81 @@ impl Qc {
...
+
+        if serializer.is_human_readable() {
+            let size = N * 2 + 2;
+            let mut s = String::with_capacity(size);
+            s.push_str("0x");
+            for v in src {
+                // unwrap is safe because we allocate enough space
+                write!(&mut s, "{:02x}", v).unwrap();
+            }
+            s.serialize(serializer)
```
Comment (danielSanchezQ) [inline consensus-engine/src/types.rs:210]: Also a function.

**Review by danielSanchezQ — CHANGES_REQUESTED** [review]

File: consensus-engine/src/utils.rs [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+        serializer: S,
+    ) -> Result<S::Ok, S::Error> {
+        if serializer.is_human_readable() {
+            serialize_human_readable_bytes_array(src, serializer)
+        } else {
+            serializer.serialize_bytes(&src)
+        }
+    }
+
+    fn deserialize_human_readable_bytes_array<'de, const N: usize, D: serde::Deserializer<'de>>(
```
Comment (zeegomo) [inline consensus-engine/src/utils.rs:30]: there has to be a crate that does this (e.g. `hex`)

**Review by zeegomo — CHANGES_REQUESTED** [review]

File: consensus-engine/src/utils.rs (hunk 5 above)
Comment (al8n) [reply consensus-engine/src/utils.rs:22]: Yeah, find a good one, switch to `const-hex`.

File: consensus-engine/src/utils.rs [hunk 6]
```diff
@@ -0,0 +1,66 @@
+#[cfg(feature = "serde")]
+pub(crate) mod serde {
```
Comment (zeegomo) [inline consensus-engine/src/utils.rs:2]: this should probably go in some more general module instead of consensus engine

**Review by zeegomo — COMMENTED** [review]
Looks good apart from the comment that I left

File: consensus-engine/src/utils.rs (hunk 6 above)
Comment (al8n) [reply consensus-engine/src/utils.rs]: Yeah, cleaner, move it to the `nomos-utils` crate.

**Review by zeegomo — APPROVED** [review]

**Pull request merged**